src/tools/singlejar/combiners.cc

```diff
@@ -14,53 +14,35 @@
 
 #include "src/tools/singlejar/combiners.h"
 
+#include <algorithm>
 #include <cctype>
+#include <cstring>
 #include <iostream>
 #include <iterator>
+#include <map>
+#include <memory>
 #include <sstream>
 #include <string>
+#include <vector>
 
 #include "src/tools/singlejar/diag.h"
 
 Combiner::~Combiner() {}
 
-Concatenator::~Concatenator() {}
-
-bool Concatenator::Merge(const CDH *cdh, const LH *lh) {
-  if (insert_newlines_ && buffer_.get() && buffer_->data_size() &&
-      '\n' != buffer_->last_byte()) {
-    Append("\n", 1);
-  }
-  CreateBuffer();
-  if (Z_NO_COMPRESSION == lh->compression_method()) {
-    buffer_->ReadEntryContents(lh);
-  } else if (Z_DEFLATED == lh->compression_method()) {
-    if (!inflater_) {
-      inflater_.reset(new Inflater());
-    }
-    buffer_->DecompressEntryContents(cdh, lh, inflater_.get());
-  } else {
-    diag_errx(2, "%s is neither stored nor deflated", filename_.c_str());
-  }
-  return true;
-}
-
-void *Concatenator::OutputEntry(bool compress) {
-  if (!buffer_) {
-    return nullptr;
-  }
-
+void *outputEntryFromBuffer(const std::string filename,
```

Comment:
Instead of extracting this helper, would it be possible to share implementation with `Concatenator` with composition, similar to what `ManifestCombiner` does?

Reply:
Composed in e416b51e12